File: src/App/wwwroot/js/BlazorMonacoInterop.js
export function executeAction(editorId, actionId) {
    window.blazorMonaco.editor.getEditor(editorId).getAction(actionId)?.run();
}

export function onDidChangeCursorPosition(editorId, callback) {
    const editor = window.blazorMonaco.editor.getEditor(editorId);
    return editor.onDidChangeCursorPosition(async (e) => {
        if (e.reason === 3 /* explicit user gesture */) {
            const model = editor.getModel();
            if (model) {
                const offset = model.getOffsetAt(e.position);
                await DotNet.invokeMethodAsync('DotNetLab.App', 'OnDidChangeCursorPositionCallbackAsync', callback, offset);
            }
        }
    });
}

export function setSelection(editorId, start, end) {
    const editor = window.blazorMonaco.editor.getEditor(editorId);
    const model = editor.getModel();
    if (model) {
        const startPosition = model.getPositionAt(start);
        const endPosition = model.getPositionAt(end);
        const range = new monaco.Range(
            startPosition.lineNumber, startPosition.column,
            endPosition.lineNumber, endPosition.column);
        editor.setSelection(range);
        editor.revealRangeInCenter(range);
    }
}

export function setModelValueUndoable(editorId, modelUri, text) {
    const editor = window.blazorMonaco.editor.getEditor(editorId);
    editor.pushUndoStop();
    const model = monaco.editor.getModel(modelUri);
    model.pushEditOperations(null, [{ range: model.getFullModelRange(), text }], () => null);
    editor.pushUndoStop();
}

/**
 * @param {string} modelUri
 * @returns {number}
 */
export function getAlternativeVersionId(modelUri) {
    const models = monaco.editor.getModels();
    const model = monaco.editor.getModel(modelUri)
        ?? models.find((candidate) => {
            const uri = candidate.uri.toString();
            return uri === modelUri || decodeURI(uri) === modelUri;
        });
    return model?.getAlternativeVersionId() ?? -1;
}

/**
 * @param {string} language
 * @param {string[] | undefined} triggerCharacters
 */
export function registerCompletionProvider(language, triggerCharacters, completionItemProvider) {
    // https://microsoft.github.io/monaco-editor/docs.html#functions/editor_editor_api.languages.registerCompletionItemProvider.html
    return monaco.languages.registerCompletionItemProvider(JSON.parse(language), {
        triggerCharacters: triggerCharacters,
        provideCompletionItems: async (model, position, context, token) => {
            const versionId = model.getAlternativeVersionId();
            const tokenRef = wrapToken(token);
            try {
                /** @type {monaco.languages.CompletionList} */
                const result = JSON.parse(await DotNet.invokeMethodAsync('DotNetLab.App', 'ProvideCompletionItemsAsync',
                    completionItemProvider, decodeURI(model.uri.toString()), JSON.stringify(position), JSON.stringify(context), tokenRef));

                if (versionId != model.getAlternativeVersionId()) {
                    ignoredProviderRequest();
                }

                for (const item of result.suggestions) {
                    // `insertText` is missing if it's equal to `label` to save bandwidth
                    // but monaco editor expects it to be always present.
                    item.insertText ??= item.label;

                    // These are the same for all completion items.
                    item.range = result.range;
                    item.commitCharacters = result.commitCharacters;
                }

                return result;
            } catch (e) {
                rethrowProviderError(e);
            } finally {
                DotNet.disposeJSObjectReference(tokenRef);
            }
        },
        resolveCompletionItem: async (completionItem, token) => {
            const tokenRef = wrapToken(token);
            try {
                const json = await DotNet.invokeMethodAsync('DotNetLab.App', 'ResolveCompletionItemAsync',
                    completionItemProvider, JSON.stringify(completionItem), tokenRef);

                if (json) {
                    const result = JSON.parse(json);
                    result.range ??= completionItem.range;
                    return result;
                }

                return completionItem;
            } catch (e) {
                rethrowProviderError(e);
            } finally {
                DotNet.disposeJSObjectReference(tokenRef);
            }
        },
    });
}

let debugSemanticTokens = false;
let semanticHighlightingHooked = false;

function applySemanticHighlighting(editor) {
    editor.updateOptions({
        'semanticHighlighting.enabled': true,
    });
    if (editor.getAction('debug-semantic-token')) {
        return;
    }

    editor.addAction({
        id: 'debug-semantic-token',
        label: 'Debug Semantic Tokens (See Browser Console)',
        run: () => {
            debugSemanticTokens = !debugSemanticTokens;
            console.log('Debugging semantic tokens ' + (debugSemanticTokens ? 'enabled' : 'disabled'));
        },
    });
}

export function enableSemanticHighlighting() {
    if (!semanticHighlightingHooked) {
        semanticHighlightingHooked = true;
        monaco.editor.onDidCreateEditor(applySemanticHighlighting);
    }

    for (const editor of window.blazorMonaco.editors.map(x => x.editor)) {
        applySemanticHighlighting(editor);
    }
}

export function registerSemanticTokensProvider(language, legend, provider, registerRangeProvider) {
    const disposables = new DisposableList();
    const languageParsed = JSON.parse(language);
    const legendParsed = JSON.parse(legend);

    // https://microsoft.github.io/monaco-editor/docs.html#functions/editor_editor_api.languages.registerDocumentSemanticTokensProvider.html
    disposables.add(monaco.languages.registerDocumentSemanticTokensProvider(languageParsed, {
        getLegend: () => legendParsed,
        provideDocumentSemanticTokens: async (model, lastResultId, token) => {
            const tokenRef = wrapToken(token);
            try {
                const result = await DotNet.invokeMethodAsync('DotNetLab.App', 'ProvideSemanticTokensAsync',
                    provider, decodeURI(model.uri.toString()), null, debugSemanticTokens, tokenRef);
                return decodeResult(result, legendParsed);
            } catch (e) {
                rethrowProviderError(e);
            } finally {
                DotNet.disposeJSObjectReference(tokenRef);
            }
        },
        releaseDocumentSemanticTokens: (resultId) => {
            // Not implemented.
        },
    }));

    if (registerRangeProvider) {
        // https://microsoft.github.io/monaco-editor/docs.html#functions/editor_editor_api.languages.registerDocumentRangeSemanticTokensProvider.html
        disposables.add(monaco.languages.registerDocumentRangeSemanticTokensProvider(languageParsed, {
            getLegend: () => legendParsed,
            provideDocumentRangeSemanticTokens: async (model, range, token) => {
                const tokenRef = wrapToken(token);
                try {
                    const result = await DotNet.invokeMethodAsync('DotNetLab.App', 'ProvideSemanticTokensAsync',
                        provider, decodeURI(model.uri.toString()), JSON.stringify(range), debugSemanticTokens, tokenRef);
                    return decodeResult(result, legendParsed);
                } catch (e) {
                    rethrowProviderError(e);
                } finally {
                    DotNet.disposeJSObjectReference(tokenRef);
                }
            },
        }));
    }

    return disposables;

    function decodeResult(result, legend) {
        if (result === null) {
            // Null means ignore this request so current tokens stay.
            // Monaco treats Error name/message "Canceled" as cancellation, not a failure.
            ignoredProviderRequest();
        }

        // Result is Base64-encoded int32 array we want to convert to Uint32Array.
        const bytes = Uint8Array.from(atob(result), c => c.charCodeAt(0));
        const data = new Uint32Array(bytes.buffer, bytes.byteOffset, bytes.length / Uint32Array.BYTES_PER_ELEMENT);

        if (debugSemanticTokens) {
            const tokenTypes = [];
            for (let i = 0; i < data.length; i += 5) {
                tokenTypes.push(legend.tokenTypes[data[i + 3]]);
            }
            console.log('Semantic tokens:', tokenTypes);
        }

        return {
            data: data,
            resultId: null, // Currently not used.
        };
    }
}

export function registerCodeActionProvider(language, codeActionProvider) {
    // https://microsoft.github.io/monaco-editor/docs.html#functions/editor_editor_api.languages.registerCodeActionProvider.html
    return monaco.languages.registerCodeActionProvider(JSON.parse(language), {
        provideCodeActions: async (model, range, context, token) => {
            const tokenRef = wrapToken(token);
            try {
                const result = JSON.parse(await DotNet.invokeMethodAsync('DotNetLab.App', 'ProvideCodeActionsAsync',
                    codeActionProvider, decodeURI(model.uri.toString()), JSON.stringify(range), tokenRef));

                if (result === null) {
                    // If null result is returned, it means the request should be ignored, so we need to throw
                    // (as opposed to returning no code actions).
                    // The text 'busy' is recommended for this purpose (e.g., it avoids sending telemetry).
                    ignoredProviderRequest();
                }

                for (const action of result) {
                    for (const edit of action.edit?.edits ?? []) {
                        edit.resource = monaco.Uri.parse(edit.resource);
                    }
                }

                return {
                    actions: result,
                    dispose: () => { }, // Currently not used.
                };
            } catch (e) {
                rethrowProviderError(e);
            } finally {
                DotNet.disposeJSObjectReference(tokenRef);
            }
        },
    }, {
        providedCodeActionKinds: ['quickfix'],
    });
}

export function registerDefinitionProvider(language, definitionProvider) {
    // https://microsoft.github.io/monaco-editor/docs.html#functions/editor_editor_api.languages.registerDefinitionProvider.html
    return monaco.languages.registerDefinitionProvider(JSON.parse(language), {
        provideDefinition: async (model, position, token) => {
            try {
                const offset = model.getOffsetAt(position);
                const result = await DotNet.invokeMethodAsync('DotNetLab.App', 'ProvideDefinition',
                    definitionProvider, decodeURI(model.uri.toString()), offset);

                if (result === null) {
                    return null;
                }

                const [start, end] = result.split(';');
                const startPosition = model.getPositionAt(start);
                const endPosition = model.getPositionAt(end);
                const range = new monaco.Range(
                    startPosition.lineNumber, startPosition.column,
                    endPosition.lineNumber, endPosition.column);
                return {
                    uri: model.uri,
                    range,
                };
            } catch (e) {
                console.error(e);
                throw e;
            }
        },
    });
}

export function registerHoverProvider(language, hoverProvider) {
    // https://microsoft.github.io/monaco-editor/docs.html#functions/editor_editor_api.languages.registerHoverProvider.html
    return monaco.languages.registerHoverProvider(JSON.parse(language), {
        provideHover: async (model, position, token, context) => {
            const tokenRef = wrapToken(token);
            try {
                const result = await DotNet.invokeMethodAsync('DotNetLab.App', 'ProvideHoverAsync',
                    hoverProvider, decodeURI(model.uri.toString()), JSON.stringify(position), tokenRef);

                if (result === null) {
                    // If null result is returned, it means the request should be ignored, so we need to throw
                    // (as opposed to returning no code actions).
                    // The text 'busy' is recommended for this purpose (e.g., it avoids sending telemetry).
                    ignoredProviderRequest();
                }

                return { contents: [{ value: result }] };
            } catch (e) {
                rethrowProviderError(e);
            } finally {
                DotNet.disposeJSObjectReference(tokenRef);
            }
        },
    });
}

export function registerSignatureHelpProvider(language, hoverProvider) {
    // https://microsoft.github.io/monaco-editor/docs.html#functions/editor_editor_api.languages.registerSignatureHelpProvider.html
    return monaco.languages.registerSignatureHelpProvider(JSON.parse(language), {
        signatureHelpTriggerCharacters: ['(', ','],
        provideSignatureHelp: async (model, position, token, context) => {
            const tokenRef = wrapToken(token);
            try {
                const contextLight = {
                    ...context,
                    // Avoid sending currently unused data.
                    activeSignatureHelp: undefined,
                };

                const result = await DotNet.invokeMethodAsync('DotNetLab.App', 'ProvideSignatureHelpAsync',
                    hoverProvider, decodeURI(model.uri.toString()), JSON.stringify(position), JSON.stringify(contextLight), tokenRef);

                if (result === null) {
                    // If null result is returned, it means the request should be ignored, so we need to throw
                    // (as opposed to returning no code actions).
                    // The text 'busy' is recommended for this purpose (e.g., it avoids sending telemetry).
                    ignoredProviderRequest();
                }

                const parsed = JSON.parse(result);

                if (parsed === null) {
                    return null;
                }

                return {
                    value: parsed,
                    dispose: () => { }, // Currently not used.
                };
            } catch (e) {
                rethrowProviderError(e);
            } finally {
                DotNet.disposeJSObjectReference(tokenRef);
            }
        },
    });
}

/**
 * @param {string} editorId
 * @param {number[]} offsets - start,end,start,end,...
 */
export function underlineLinks(editorId, offsets) {
    const editor = window.blazorMonaco.editor.getEditor(editorId);
    const model = editor.getModel();
    if (model) {
        const ranges = [];
        for (let i = 0; i < offsets.length; i += 2) {
            const startPosition = model.getPositionAt(offsets[i]);
            const endPosition = model.getPositionAt(offsets[i + 1]);
            const range = new monaco.Range(
                startPosition.lineNumber, startPosition.column,
                endPosition.lineNumber, endPosition.column);
            ranges.push(range);
        }

        editor.createDecorationsCollection(ranges.map(range => ({
            range,
            options: { inlineClassName: 'underline' },
        })));
    }
}

export function registerLanguage(languageId) {
    monaco.languages.register({ id: languageId });
    monaco.languages.setLanguageConfiguration(languageId, {
        folding: {
            offSide: true,
        },
    });
}

export function hasDarkTheme(editorId) {
    const editor = window.blazorMonaco.editor.getEditor(editorId);
    const theme = editor.getRawOptions().theme;
    return Boolean(theme) && theme.includes('dark');
}

function wrapToken(token) {
    return DotNet.createJSObjectReference({ 
        onCancellationRequested(tokenWrapper) {
            token.onCancellationRequested(async () => {
                await tokenWrapper.invokeMethodAsync('Cancel');
            });
        },
    });
}

function ignoredProviderRequest() {
    const error = new Error('Canceled');
    error.name = 'Canceled';
    throw error;
}

function rethrowProviderError(e) {
    if (e instanceof Error && (e.message === 'busy' || e.name === 'Canceled')) {
        throw e;
    }
    console.error(e);
    throw e;
}

class DisposableList {
    constructor() {
        this.disposables = [];
    }
    add(disposable) {
        this.disposables.push(disposable);
    }
    dispose() {
        for (const disposable of this.disposables) {
            disposable.dispose();
        }
        this.disposables = [];
    }
}

File: src/App/Editor/Monaco/BlazorMonacoInterop.cs
using BlazorMonaco.Languages;
using Microsoft.Extensions.Logging;
using Microsoft.JSInterop;
using System.Runtime.CompilerServices;
using System.Text.Json;

namespace DotNetLab.Editor.Monaco;

using CursorPositionCallback = Func<int, Task>;

/// <remarks>
/// Needs to be public for <see cref="JSInvokableAttribute"/>.
/// </remarks>
public sealed partial class BlazorMonacoInterop : IAsyncDisposable
{
    private const string moduleName = nameof(BlazorMonacoInterop);

    private readonly Lazy<Task<IJSObjectReference>> initialize;
    private readonly ILogger<BlazorMonacoInterop> _logger;

    public BlazorMonacoInterop(IJSRuntime jsRuntime, ILogger<BlazorMonacoInterop> logger)
    {
        _logger = logger;
        initialize = new(() => jsRuntime.InvokeAsync<IJSObjectReference>("import", "../_content/DotNetLab.App/js/BlazorMonacoInterop.js").AsTask());
    }

    public async ValueTask DisposeAsync()
    {
        if (initialize.IsValueCreated)
        {
            await (await Module).DisposeAsync();
        }
    }

    private Task<IJSObjectReference> Module => initialize.Value;

    [JSInvokable]
    public static async Task OnDidChangeCursorPositionCallbackAsync(
        DotNetObjectReference<CursorPositionCallback> callback, int offset)
    {
        var func = callback.Value;
        await func(offset);
    }

    [JSInvokable]
    public static async Task<string> ProvideCompletionItemsAsync(
        DotNetObjectReference<CompletionItemProviderAsync> completionItemProviderReference,
        string modelUri,
        string position,
        string context,
        IJSObjectReference token)
    {
        var completionItemProvider = completionItemProviderReference.Value;
        using var tokenWrapper = await ToCancellationTokenAsync(token, completionItemProvider.Logger);
        string json = await completionItemProvider.ProvideCompletionItemsAsync(
            modelUri,
            JsonSerializer.Deserialize(position, BlazorMonacoJsonContext.Default.Position)!,
            JsonSerializer.Deserialize(context, BlazorMonacoJsonContext.Default.CompletionContext)!,
            tokenWrapper.Token);
        return json.Length > 64 * 1024
            ? """{"suggestions":[]}"""
            : json;
    }

    [JSInvokable]
    public static async Task<string?> ResolveCompletionItemAsync(
        DotNetObjectReference<CompletionItemProviderAsync> completionItemProviderReference,
        string item,
        IJSObjectReference token)
    {
        var completionItemProvider = completionItemProviderReference.Value;
        using var tokenWrapper = await ToCancellationTokenAsync(token, completionItemProvider.Logger);
        string? json = await completionItemProvider.ResolveCompletionItemAsync(
            JsonSerializer.Deserialize(item, BlazorMonacoJsonContext.Default.MonacoCompletionItem)!,
            tokenWrapper.Token);
        return json;
    }

    [JSInvokable]
    public static async Task<string?> ProvideSemanticTokensAsync(
        DotNetObjectReference<SemanticTokensProvider> providerReference,
        string modelUri,
        string? rangeJson,
        bool debug,
        IJSObjectReference token)
    {
        var provider = providerReference.Value;
        using var tokenWrapper = await ToCancellationTokenAsync(token, provider.Logger);
        string? json = await provider.ProvideSemanticTokens(modelUri, rangeJson, debug, tokenWrapper.Token);
        return json;
    }

    [JSInvokable]
    public static async Task<string?> ProvideCodeActionsAsync(
        DotNetObjectReference<CodeActionProviderAsync> providerReference,
        string modelUri,
        string? rangeJson,
        IJSObjectReference token)
    {
        var provider = providerReference.Value;
        using var tokenWrapper = await ToCancellationTokenAsync(token, provider.Logger);
        string? json = await provider.ProvideCodeActions(modelUri, rangeJson, tokenWrapper.Token);
        return json;
    }

    [JSInvokable]
    public static string? ProvideDefinition(
        DotNetObjectReference<DefinitionProvider> providerReference,
        string modelUri,
        int offset)
    {
        var provider = providerReference.Value;
        var span = provider.ProvideDefinition(modelUri, offset);
        return span is { } value ? $"{value.Start};{value.End}" : null;
    }

    [JSInvokable]
    public static async Task<string?> ProvideHoverAsync(
        DotNetObjectReference<HoverProvider> providerReference,
        string modelUri,
        string positionJson,
        IJSObjectReference token)
    {
        var provider = providerReference.Value;
        using var tokenWrapper = await ToCancellationTokenAsync(token, provider.Logger);
        string? json = await provider.ProvideHover(modelUri, positionJson, tokenWrapper.Token);
        return json;
    }

    [JSInvokable]
    public static async Task<string?> ProvideSignatureHelpAsync(
        DotNetObjectReference<SignatureHelpProvider> providerReference,
        string modelUri,
        string positionJson,
        string contextJson,
        IJSObjectReference token)
    {
        var provider = providerReference.Value;
        using var tokenWrapper = await ToCancellationTokenAsync(token, provider.Logger);
        string? json = await provider.ProvideSignatureHelp(modelUri, positionJson, contextJson, tokenWrapper.Token);
        return json;
    }

    public async Task ExecuteActionAsync(string editorId, string actionId)
    {
        await (await Module).InvokeVoidAsync("executeAction", editorId, actionId);
    }

    public async Task<IAsyncDisposable> OnDidChangeCursorPositionAsync(string editorId, CursorPositionCallback callback)
    {
        var callbackRef = DotNetObjectReference.Create(callback);
        var disposable = await (await Module).InvokeAsync<IJSObjectReference>("onDidChangeCursorPosition", editorId, callbackRef);
        return new Disposable(disposable, callbackRef);
    }

    public async Task SetSelectionAsync(string editorId, int start, int end)
    {
        await (await Module).InvokeVoidAsync("setSelection", editorId, start, end);
    }

    public async Task SetModelValueUndoable(string editorId, string modelUri, string text)
    {
        await (await Module).InvokeVoidAsync("setModelValueUndoable", editorId, modelUri, text);
    }

    public async Task<int> GetAlternativeVersionIdAsync(string modelUri)
    {
        try
        {
            return await (await Module).InvokeAsync<int>("getAlternativeVersionId", modelUri);
        }
        catch (JSException ex)
        {
            _logger.LogDebug(ex, "Reading the alternative model version failed.");
            return -1;
        }
    }

    public async Task<IAsyncDisposable> RegisterCompletionProviderAsync(
        LanguageSelector language,
        CompletionItemProviderAsync completionItemProvider)
    {
        var completionItemProviderRef = DotNetObjectReference.Create(completionItemProvider);
        var disposable = await (await Module).InvokeAsync<IJSObjectReference>("registerCompletionProvider",
            JsonSerializer.Serialize(language, BlazorMonacoJsonContext.Default.LanguageSelector),
            completionItemProvider.TriggerCharacters,
            completionItemProviderRef);
        return new Disposable(disposable, completionItemProviderRef);
    }

    public async Task EnableSemanticHighlightingAsync()
    {
        await (await Module).InvokeVoidAsync("enableSemanticHighlighting");
    }

    public async Task<IAsyncDisposable> RegisterSemanticTokensProviderAsync(
        LanguageSelector language,
        SemanticTokensProvider provider)
    {
        var providerRef = DotNetObjectReference.Create(provider);
        var disposable = await (await Module).InvokeAsync<IJSObjectReference>("registerSemanticTokensProvider",
            JsonSerializer.Serialize(language, BlazorMonacoJsonContext.Default.LanguageSelector),
            JsonSerializer.Serialize(provider.Legend, BlazorMonacoJsonContext.Default.SemanticTokensLegend),
            providerRef,
            provider.RegisterRangeProvider);
        return new Disposable(disposable, providerRef);
    }

    public async Task<IAsyncDisposable> RegisterCodeActionProviderAsync(
        LanguageSelector language,
        CodeActionProviderAsync provider)
    {
        var providerRef = DotNetObjectReference.Create(provider);
        var disposable = await (await Module).InvokeAsync<IJSObjectReference>("registerCodeActionProvider",
            JsonSerializer.Serialize(language, BlazorMonacoJsonContext.Default.LanguageSelector),
            providerRef);
        return new Disposable(disposable, providerRef);
    }

    public async Task<IAsyncDisposable> RegisterDefinitionProviderAsync(
        LanguageSelector language,
        DefinitionProvider provider)
    {
        var providerRef = DotNetObjectReference.Create(provider);
        var disposable = await (await Module).InvokeAsync<IJSObjectReference>("registerDefinitionProvider",
            JsonSerializer.Serialize(language, BlazorMonacoJsonContext.Default.LanguageSelector),
            providerRef);
        return new Disposable(disposable, providerRef);
    }

    public async Task<IAsyncDisposable> RegisterHoverProviderAsync(
        LanguageSelector language,
        HoverProvider provider)
    {
        var providerRef = DotNetObjectReference.Create(provider);
        var disposable = await (await Module).InvokeAsync<IJSObjectReference>("registerHoverProvider",
            JsonSerializer.Serialize(language, BlazorMonacoJsonContext.Default.LanguageSelector),
            providerRef);
        return new Disposable(disposable, providerRef);
    }

    public async Task<IAsyncDisposable> RegisterSignatureHelpProviderAsync(
        LanguageSelector language,
        SignatureHelpProvider provider)
    {
        var providerRef = DotNetObjectReference.Create(provider);
        var disposable = await (await Module).InvokeAsync<IJSObjectReference>("registerSignatureHelpProvider",
            JsonSerializer.Serialize(language, BlazorMonacoJsonContext.Default.LanguageSelector),
            providerRef);
        return new Disposable(disposable, providerRef);
    }

    public async Task UnderlineLinksAsync(string editorId, int[] offsets)
    {
        await (await Module).InvokeVoidAsync("underlineLinks", editorId, offsets);
    }

    public async Task RegisterLanguageAsync(string languageId)
    {
        await (await Module).InvokeVoidAsync("registerLanguage", languageId);
    }

    public async Task RegisterDiagnosticsLanguageAsync(string languageId)
    {
        await (await Module).InvokeVoidAsync("registerDiagnosticsLanguage", languageId);
    }

    public async Task<bool> HasDarkThemeAsync(string editorId)
    {
        return await (await Module).InvokeAsync<bool>("hasDarkTheme", editorId);
    }

    private static async Task<CancellationTokenWrapper> ToCancellationTokenAsync(IJSObjectReference token, ILogger logger, [CallerMemberName] string memberName = "")
    {
        var tokenWrapper = new CancellationTokenWrapper();
        tokenWrapper.Token.Register(() => logger.LogDebug("Cancelling '{Member}'.", memberName));
        await token.InvokeVoidAsync("onCancellationRequested", tokenWrapper.ObjectReference);
        return tokenWrapper;
    }

    private sealed class Disposable(IJSObjectReference disposable, IDisposable? other) : IAsyncDisposable
    {
        public async ValueTask DisposeAsync()
        {
            await disposable.InvokeVoidAsync("dispose");
            await disposable.DisposeAsync();
            other?.Dispose();
        }
    }
}

public sealed class CancellationTokenWrapper
    : IDisposable
{
    private readonly CancellationTokenSource cts;
    private readonly DotNetObjectReference<CancellationTokenWrapper> objectRef;

    public CancellationTokenWrapper()
    {
        cts = new CancellationTokenSource();
        objectRef = DotNetObjectReference.Create(this);
    }

    public CancellationToken Token => cts.Token;
    public DotNetObjectReference<CancellationTokenWrapper> ObjectReference => objectRef;

    [JSInvokable]
    public void Cancel() => cts.Cancel();

    public void Dispose()
    {
        cts.Dispose();
        objectRef.Dispose();
    }
}
